### PR TITLE
Add disable instance discovery

### LINKED ISSUE #27888: Collaborate on solution for validate_authority with MSAL team
Tracking issue for collaborating with the MSAL team to expose consistent capabilities / APIs such that Azure Identity SDK can then expose a simple and easy to use API for enabling the use of custom domains.

Related Issues: 
- expose "validate_authority" so user can explicitly disable the validation if they want to use custom domain. https://github.com/Azure/azure-sdk-for-python/pull/22786/
- [ClientSecretCredential.get_token() fails in case of azurestack if we provide 'authority' URL · Issue #22625 · Azure/azure-sdk-for-python (github.com)
](https://github.com/Azure/azure-sdk-for-python/issues/22625)

**MSAL Tracking Issues:**
  - Python: https://github.com/AzureAD/microsoft-authentication-library-for-python/issues/466
  - JS: https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/4625
  - Java: https://github.com/AzureAD/microsoft-authentication-library-for-java/issues/481
  - .NET: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3242
  - Golang: https://github.com/AzureAD/microsoft-authentication-library-for-go/issues/301

### PR DESCRIPTION
**Pull request by billwert**
Add method for disabling AAD instance discovery.

Fixes #27888 

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java [hunk 1]
```diff
@@ -654,6 +656,23 @@ IdentityClientOptions setPerRetryPolicies(List<HttpPipelinePolicy> perRetryPolic
         return this;
     }
 
+    /**
+     * Disable instance discovery.
+     * @return the updated client options
+     */
+    public IdentityClientOptions disableInstanceDisovery() {
```
Comment (alzimmermsft) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:663]: I know the naming and usage here is more convenient, but I've never been a fan of this style as its not revertible, so if you can this you can never re-enable instance discovery.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/AadCredentialBuilderBase.java [hunk 2]
```diff
@@ -107,4 +107,16 @@ public T additionallyAllowedTenants(List<String> additionallyAllowedTenants) {
...
     }
+
+    /**
+     * Disables instance discovery.
+     *
+     * @return An updated instance of this builder with instance discovery disabled.
+     */
+    @SuppressWarnings("unchecked")
+
+    public T disableInstanceDiscovery() {
```
Comment (xiangyan99) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/AadCredentialBuilderBase.java:118]: Our argument name is instance_discovery which defaults to None (use whatever msal has)

If user wants to opt out, they pass in instance_discovery=False.

Comment (billwert) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/AadCredentialBuilderBase.java:118]: Understood. This is how we've represented similar concepts in Java. See also [`DeviceCodeCredentialBuilder.disableAutomaticAuthentication`](https://github.com/Azure/azure-sdk-for-java/blob/ae606de127c9aeeedd7ee8e0d68bf141f32e3ff9/sdk/identity/azure-identity/src/main/java/com/azure/identity/DeviceCodeCredentialBuilder.java#L74).

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java (hunk 1 above)
Comment (billwert) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:663]: Modifying a builder to produce a second credential seems unlikely. At this point I'm going to leave this as is for consistency with other on-by-default options in the Identity SDK.

**Review by alzimmermsft — APPROVED** [review]

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java [hunk 3]
```diff
@@ -68,6 +68,7 @@ public final class IdentityClientOptions implements Cloneable {
     private RetryPolicy retryPolicy;
     private List<HttpPipelinePolicy> perCallPolicies;
     private List<HttpPipelinePolicy> perRetryPolicies;
+    private boolean instanceDiscovery;
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:71]: include updating this field as part of clone method in this class.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/AadCredentialBuilderBase.java [hunk 4]
```diff
@@ -107,4 +107,16 @@ public T additionallyAllowedTenants(List<String> additionallyAllowedTenants) {
         identityClientOptions.setAdditionallyAllowedTenants(IdentityUtil.resolveAdditionalTenants(additionallyAllowedTenants));
         return (T) this;
     }
+
+    /**
+     * Disables instance discovery.
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/AadCredentialBuilderBase.java:112]: The Javadoc can have more detail as to what process this step actually details.
Can copy from other langauges or msal4j doc, if they have more detail there.

**Pull request merged**